front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.java

```diff
@@ -0,0 +1,287 @@
+import com.netflix.spinnaker.front50.validator.GenericValidationErrors;
+import com.netflix.spinnaker.front50.validator.PipelineValidator;
+import com.netflix.spinnaker.kork.web.exceptions.ValidationException;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.support.DefaultMessageSourceResolvable;
+import org.springframework.security.access.prepost.PostAuthorize;
+import org.springframework.security.access.prepost.PostFilter;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+
+/** Controller for presets */
+@RestController
+@RequestMapping("pipelines")
+public class PipelineController {
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
+  private final PipelineDAO pipelineDAO;
+  private final ObjectMapper objectMapper;
+  private final Optional<ServiceAccountsService> serviceAccountsService;
+  private final List<PipelineValidator> pipelineValidators;
+  private final Optional<PipelineTemplateDAO> pipelineTemplateDAO;
+
+  public PipelineController(
+      PipelineDAO pipelineDAO,
+      ObjectMapper objectMapper,
+      Optional<ServiceAccountsService> serviceAccountsService,
+      List<PipelineValidator> pipelineValidators,
+      Optional<PipelineTemplateDAO> pipelineTemplateDAO) {
+    this.pipelineDAO = pipelineDAO;
+    this.objectMapper = objectMapper;
+    this.serviceAccountsService = serviceAccountsService;
+    this.pipelineValidators = pipelineValidators;
+    this.pipelineTemplateDAO = pipelineTemplateDAO;
+  }
+
+  @PreAuthorize("#restricted ? @fiatPermissionEvaluator.storeWholePermission() : true")
+  @PostFilter("#restricted ? hasPermission(filterObject.name, 'APPLICATION', 'READ') : true")
+  @RequestMapping(value = "", method = RequestMethod.GET)
+  public Collection<Pipeline> list(
+      @RequestParam(required = false, value = "restricted", defaultValue = "true")
+          boolean restricted,
+      @RequestParam(required = false, value = "refresh", defaultValue = "true") boolean refresh) {
+    return pipelineDAO.all(refresh);
+  }
+
+  @PreAuthorize("hasPermission(#application, 'APPLICATION', 'READ')")
+  @RequestMapping(value = "{application:.+}", method = RequestMethod.GET)
+  public List<Pipeline> listByApplication(
+      @PathVariable(value = "application") String application,
+      @RequestParam(required = false, value = "refresh", defaultValue = "true") boolean refresh) {
+    List<Pipeline> pipelines =
+        (List<Pipeline>) pipelineDAO.getPipelinesByApplication(application, refresh);
+
+    pipelines.sort(
+        (p1, p2) -> {
+          if (p1.getIndex() != null && p2.getIndex() == null) {
+            return -1;
+          }
+          if (p1.getIndex() == null && p2.getIndex() != null) {
+            return 1;
+          }
+          if (!p1.getIndex().equals(p2.getIndex())) {
```

Comment:
I think this introduces a new NPE if both indexes are null.

Reply:
Ha. Right you are. I guess this NPE has existed since... 2014. Will fix, good catch.